### DISCUSSION TITLE
模型的输出不需要经过softmax吗

### DISCUSSION THREAD

**Discussion by mumu029** [jingyaogong/minimind#294] [Q&A]
我看模型最后一层的激活值，经过全连接后就直接作为logits输出了，这里不需要做一个softmax吗？

**Comment by basil-k-aji-dev** [comment]
softmax 并没有被漏掉，它是在**需要概率分布的地方**才做的，而在训练那条路上是**故意不做**的。

### 训练时：不能做，做了就错了

`model/model_minimind.py:257`：

```python
loss = F.cross_entropy(x.view(-1, x.size(-1)), y.view(-1), ignore_index=-100)
```

PyTorch 的 `F.cross_entropy` 内部本身就是 `log_softmax + nll_loss`，它要求输入是**原始 logits**。如果在传进去之前先做一次 softmax，就等于做了两次：`log_softmax(softmax(x))`。结果不是「多此一举」那么轻，而是分布被压平、梯度变小，训练会明显变慢甚至学不动。

所以这里保持 `lm_head` 的原始输出，是必须的，不是省略。

### 推理时：做了，就在采样那一步

同一个文件的 `generate()`，`model_minimind.py:283`：

```python
next_token = torch.multinomial(torch.softmax(logits, dim=-1), num_samples=1) if do_sample else torch.argmax(logits, dim=-1, keepdim=True)
```

`torch.multinomial` 需要的就是概率，所以这里显式做了 `softmax`。

top-p 截断那一步也做了，`model_minimind.py:280`：

```python
mask = torch.cumsum(torch.softmax(sorted_logits, dim=-1), dim=-1) > top_p
```

因为累积概率必须先归一化才有意义。

### 贪心解码为什么可以不做

`do_sample=False` 时走的是 `torch.argmax(logits, ...)`，不需要 softmax —— softmax 是**单调递增**的，不改变元素之间的大小顺序，所以 `argmax(softmax(x)) == argmax(x)`。多算一次只是浪费。

同理，`logits / temperature`（第 272 行）和 top-k 的 `-inf` 掩码（第 277-278 行）也都在 logits 空间做，因为这些操作在归一化之前做才是对的。

---

一句话：logits 是「未归一化的分数」，softmax 只在要把它当成概率用的时候才加 —— 采样、top-p 需要，交叉熵自带，argmax 不需要。

以上基于当前 main（`cc312c1`）。@jingyaogong 麻烦确认下。
